pkg/graveler/ref/commit_iterator.go

```diff
@@ -116,41 +119,63 @@ func (ci *CommitIterator) Next() bool {
 		}
 	}
 
-	if ci.queue.Len() == 0 {
-		ci.value = nil
-		ci.state = commitIteratorStateDone
-		return false
-	}
+	for ci.queue.Len() > 0 {
+		ci.value = heap.Pop(&ci.queue).(*graveler.CommitRecord)
 
-	// as long as we have something in the queue we will
-	// set it as the current value and push the current commits parents to the queue
-	ci.value = heap.Pop(&ci.queue).(*graveler.CommitRecord)
-	parents := ci.value.Parents
-	if ci.firstParent && len(parents) > 1 {
-		parents = parents[:1]
-	}
-	for _, p := range parents {
-		// skip commits we already visited
-		if _, visited := ci.visit[p]; visited {
+		if ci.noMerges && len(ci.value.Parents) > 1 {
+			firstParent := ci.value.Parents[0]
+
+			if _, visited := ci.visit[firstParent]; visited {
+				continue
+			}
+
+			rec, err := ci.getCommitRecord(firstParent)
+			if err != nil {
+				ci.value = nil
+				ci.err = err
+				return false
+			}
+			ci.visit[rec.CommitID] = struct{}{}
+
+			if ci.since != nil && rec.Commit.CreationDate.Before(*ci.since) {
+				continue
+			}
+			heap.Push(&ci.queue, rec)
 			continue
 		}
 
-		rec, err := ci.getCommitRecord(p)
-		if err != nil {
-			ci.value = nil
-			ci.err = err
-			return false
+		parents := ci.value.Parents
+		if ci.firstParent && len(parents) > 1 {
+			parents = parents[:1]
 		}
-		ci.visit[rec.CommitID] = struct{}{}
-
-		// skip commits that are older than since time
-		if ci.since != nil && rec.Commit.CreationDate.Before(*ci.since) {
-			continue
+		for _, p := range parents {
+			if _, visited := ci.visit[p]; visited {
+				continue
+			}
+
+			rec, err := ci.getCommitRecord(p)
+			if err != nil {
+				ci.value = nil
+				ci.err = err
+				return false
+			}
+			ci.visit[rec.CommitID] = struct{}{}
+
+			// skip commits that are older than since time
+			if ci.since != nil && rec.Commit.CreationDate.Before(*ci.since) {
+				continue
+			}
+
+			heap.Push(&ci.queue, rec)
 		}
 
-		heap.Push(&ci.queue, rec)
+		return true
```

Comment:
Impressive - One of the more delicate areas in Graveler and it seems really thought through.

For me, as a reviewer, it's near impossible to follow the logic that is applied here without comments walking me through the decisions taken in every branch of this code.